### ISSUE TITLE
Helidon DB Client

### ISSUE THREAD

**Issue by tomas-langer** [oracle/helidon#581]
This is a tracking issue for Helidon DB API and implementation.

Overall the implementation should follow the Development guidelines and:

- The implementation is full reactive and nonblocking
- There should be no dependency of the API or core on Tracing (not even Opentracing API)
- There should be no dependency of the API or core on Metrics
- There should be no dependency of the API or core on Healthchecks
- Rest will be defined in One pager

Tasks:

- [x] One Pager
- [x] API design
- [x] Core implementation
- [x] Implementation over JDBC
- [x] Metrics support (library + example)
- [x] Tracing support (library + example)
- [x] Health check support (library + example)
- [x] Implementation over Mongo reactive driver

Documentation moved to documentation epic: #1515 
Native image moved to native image SE epic: #1656

**Comment by tomas-langer** [comment]
Depends on #596

**Comment by tomas-langer** [comment]
Depends on #473

**Comment by tomas-langer** [comment]
# TODOs:
## All
 - [ ] Unit tests...
 - [ ] The code is now an advanced POC - must revisit most of the code to make sure it is production quality
 - [ ] Statement parsing (when using named params in JDBC and any params in Mongo) - the usage of a state machine is very complicated. Could we maybe live with a regular expression and limitations on what can be in the statement (such as "no comments allowed in statements")? If not, The implementation must be done again and according to our code guidelines (and also a lot lot more readable)
## Core and Common
 - [ ] Find a good way to stream results to web server (see `DbResultSupport` in both examples)
 - [ ] Make sure all error states are correctly propagated to interceptors and to user
## JDBC
- [ ] Use a connection pool (see `ConnectionPool` interface, line 90) - and add appropriate documentation of configurable options
- [ ] Implement support for mapping of `DbRowResult` (see mongoDB impl)
- [ ] Document configuration properties of the `JdbcDbProviderBuilder`
- [ ] Make sure the `Flow` API is correctly implemented

## Mongo
- [ ] In `MongoDbStatementDml` - see TODO
- [ ] Make sure the backpressure is correctly implemented in `MongoDbStatementQuery`

**Comment by tomas-langer** [comment]
DB was not chosen wisely - it looks like a Database implementation.
We are creating a "Reactive DB Client" so we should rename the modules, packages and classes to reflect that

**Comment by spericas** [comment]
@tomas-langer Yes. I'd say "Reactive DB API" or "Reactive DB Client API". To me the Client is what you write for each DB itself.

**Comment by Tomas-Kraus** [comment]
Most of the tasks are already done. Adding few more things that came to my mind:

- [ ] **Start JDBC client with offline database**
MongoDB client already supports this and there are real-life use-cases for such a feature.
Better health check support when DB is offline.

- [ ] **Connection Pool in Metrics**
User needs access to CP settings and also current CP state attributes.
See https://github.com/brettwooldridge/HikariCP/wiki/Dropwizard-Metrics for implementation details

- [ ] **Update samples to cover typical use-cases**
